src/Symfony/Component/HttpClient/Response/HttplugPromise.php

```diff
@@ -0,0 +1,236 @@
+        $clientPool = self::$promisePool[$client] ?? self::$promisePool[$client] = new \SplObjectStorage();
+        $clientPool[$response] = $this;
+    }
+
+    public function cancel(): void
+    {
+        if (Promise::PENDING !== $this->state) {
+            return;
+        }
+
+        $this->callbacks = [];
+
+        if ($this->response) {
+            $this->response->cancel();
+        } elseif ($this->result instanceof self || $this->result instanceof PromiseInterface) {
+            $this->result->cancel();
+        }
+
+        $this->wait(false);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function then(callable $onFulfilled = null, callable $onRejected = null): Promise
+    {
+        if ($onFulfilled || $onRejected) {
+            $this->callbacks[] = [$onFulfilled, $onRejected];
+        }
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @return Psr7ResponseInterface|mixed
+     */
+    public function wait($unwrap = true)
+    {
+        $responseStream = $promisePool = [];
+        if ($this->response) {
+            $promisePool = self::$promisePool[$this->client];
+            $responseStream = $this->client->stream($promisePool);
+        }
+
+        foreach ($responseStream as $response => $chunk) {
+            try {
+                if ($chunk->isFirst()) {
+                    // Deactivate throwing on 3/4/5xx
+                    $response->getStatusCode();
+                } elseif ($chunk->isLast() && $promise = $promisePool[$response] ?? null) {
+                    // Cast the Symfony response into a PSR-7 one
+                    $psrResponse = $this->responseFactory->createResponse($response->getStatusCode());
+
+                    foreach ($response->getHeaders(false) as $name => $values) {
+                        foreach ($values as $value) {
+                            $psrResponse = $psrResponse->withAddedHeader($name, $value);
+                        }
+                    }
+
+                    $body = isset(class_uses($response)[ResponseTrait::class]) ? $response->toStream(false) : StreamWrapper::createResource($response, $this->client);
+                    $body = $this->streamFactory->createStreamFromResource($body);
+
+                    if ($body->isSeekable()) {
+                        $body->seek(0);
+                    }
+
+                    $psrResponse = $psrResponse->withBody($body);
+                    unset($promisePool[$response]);
+                    $promise->fulfill($psrResponse);
+                }
+            } catch (\Exception $e) {
+                if ($e instanceof TransportExceptionInterface) {
+                    $e = new NetworkException($e->getMessage(), $this->request, $e);
+                }
+
+                $promise = $promisePool[$response];
+                unset($promisePool[$response]);
+                $promise->reject($e);
+            }
+
+            if (!$this->response) {
```

Comment:
This line is a bit out of context. We should add a comment saying: if this promise just got fulfilled.

Reply:
added